File: Sample Code/ObjcSampleCode/Demo/DJISdkDemo/Demo/FlightController/Compass/FCCompassViewController.m
//
//  FCCompassViewController.m
//  DJISdkDemo
//
//  Copyright © 2015 DJI. All rights reserved.
//
/**
 *  This file demonstrates how to use compass calibration of DJIFlightController. 
 */
#import "FCCompassViewController.h"
#import "DemoComponentHelper.h"
#import "DemoAlertView.h"
#import <DJISDK/DJISDK.h>

@interface FCCompassViewController () <DJIFlightControllerDelegate>

@property (weak, nonatomic) IBOutlet UILabel *headingLabel;
@property (weak, nonatomic) IBOutlet UILabel *calibratingLabel;
@property (weak, nonatomic) IBOutlet UILabel *statusLabel;

- (IBAction)onCompassCalibrationButtonClicked:(id)sender;
@end

@implementation FCCompassViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view from its nib.
    
    DJIFlightController* fc = [DemoComponentHelper fetchFlightController];
    if (fc) {
        fc.delegate = self;
    }
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (IBAction)onCompassCalibrationButtonClicked:(UIButton*)sender {
    const int STOP_TAG = 100;
    const int START_TAG = 101;
    DJIFlightController* fc = [DemoComponentHelper fetchFlightController];
    if (fc) {
        if (sender.tag == STOP_TAG) {
            [fc.compass stopCalibrationWithCompletion:^(NSError * _Nullable error) {
                if (error) {
                    ShowResult(@"Stop Calibration:%@", error.localizedDescription);
                }
                else
                {
                    [sender setTitle:@"Start Calibration" forState:UIControlStateNormal];
                    sender.tag = START_TAG;
                }
            }];
        }
        else
        {
            [fc.compass startCalibrationWithCompletion:^(NSError * _Nullable error) {
                if (error) {
                    ShowResult(@"Start Calibration:%@", error.localizedDescription);
                }
                else
                {
                    [sender setTitle:@"Stop Calibration" forState:UIControlStateNormal];
                    sender.tag = STOP_TAG;
                }
            }];
        }
    }
    else
    {
        ShowResult(@"Component Not Exist.");
    }
}

- (void)flightController:(DJIFlightController *)fc didUpdateSystemState:(DJIFlightControllerCurrentState *)state
{
    self.headingLabel.text = [NSString stringWithFormat:@"%0.1f", fc.compass.heading];
    self.calibratingLabel.text = fc.compass.isCalibrating ? @"YES" : @"NO";
    self.statusLabel.text = [self stringWithCalibrationStatus:fc.compass.calibrationStatus];
}

-(NSString*) stringWithCalibrationStatus:(DJICompassCalibrationStatus)status
{
    if (status == DJICompassCalibrationStatusNone) {
        return @"None";
    }
    else if (status == DJICompassCalibrationStatusHorizontal)
    {
        return @"Horizontal";
    }
    else if (status == DJICompassCalibrationStatusVertical)
    {
        return @"Vertical";
    }
    else if (status == DJICompassCalibrationStatusSucceeded)
    {
        return @"Succeeded";
    }
    else if (status == DJICompassCalibrationStatusFailed)
    {
        return @"Failed";
    }
    else
    {
        return @"Unknown";
    }
}
@end

File: Sample Code/SwiftSampleCode/DJISDKSwiftDemo/Components/FlightController/Compass/FCCompassViewController.swift
//
//  FCCompassViewController.swift
//  DJISDKSwiftDemo
//
//  Created by DJI on 16/1/5.
//  Copyright © 2016 DJI. All rights reserved.
//

import DJISDK
class FCCompassViewController: DJIBaseViewController, DJIFlightControllerDelegate {
    
    @IBOutlet weak var headingLabel: UILabel!
    @IBOutlet weak var calibratingLabel: UILabel!
    @IBOutlet weak var statusLabel: UILabel!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view from its nib.
        let fc: DJIFlightController? = self.fetchFlightController()
        if fc != nil {
            fc!.delegate = self
        }
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    @IBAction func onCompassCalibrationButtonClicked(_ sender: UIButton) {
        let STOP_TAG: Int = 100
        let START_TAG: Int = 101
        let fc: DJIFlightController? = self.fetchFlightController()
        if fc != nil {
            if sender.tag == STOP_TAG {
                fc?.compass?.stopCalibration(completion: {[weak self](error: Error?) -> Void in
                    if error != nil {
                        self?.showAlertResult("Stop Calibration:\(error?.localizedDescription)")
                    }
                    else {
                        sender.setTitle("Start Calibration", for: UIControlState())
                        sender.tag = START_TAG
                    }
                })
            }
            else {
                fc?.compass?.startCalibration(completion: {[weak self](error: Error?) -> Void in
                    if error != nil {
                        self?.showAlertResult("Start Calibration:\(error?.localizedDescription)")
                    }
                    else {
                        sender.setTitle("Stop Calibration", for: UIControlState())
                        sender.tag = STOP_TAG
                    }
                })
            }
        }
        else {
            self.showAlertResult("Component Not Exist.")
        }
    }
    
    func flightController(_ fc: DJIFlightController, didUpdateSystemState state: DJIFlightControllerCurrentState)
    {
        if (fc.compass == nil) {
            return;
        }
        
        self.headingLabel.text = String(format: "%0.1f", fc.compass!.heading)
        self.calibratingLabel.text = fc.compass!.isCalibrating ? "YES" : "NO"
        self.statusLabel.text = self.stringWithCalibrationStatus(fc.compass!.calibrationStatus)
    }
    
    func stringWithCalibrationStatus(_ status: DJICompassCalibrationStatus) -> String {
        if status == DJICompassCalibrationStatus.none {
            return "None"
        }
        else if status == DJICompassCalibrationStatus.horizontal {
            return "Horizontal"
        }
        else if status == DJICompassCalibrationStatus.vertical {
            return "Vertical"
        }
        else if status == DJICompassCalibrationStatus.succeeded {
            return "Succeeded"
        }
        else if status == DJICompassCalibrationStatus.failed {
            return "Failed"
        }
        else {
            return "Unknown"
        }
        
    }
    
}
